Add Find Courses and Course Info functionalities

abhra-clarice opened the pull request:
JIRA: https://openedx.atlassian.net/browse/MOB-1416

<ul>
<li>Added Find Courses screen</li>
<li>Added Course Info screen</li>
<li>Developed logic for URL parsing and conditional loading</li>
</ul>

CC @aleffert

aleffert commented on edXVideoLocker/OEXConfig.h:
```diff
@@ -46,4 +46,7 @@
 - (NSString*)newRelicKey;
 
+//Enrollment
+- (NSDictionary *)courseEnrollmentProperties;
```
Please make a full class for this, like the ``EnrollmentConfig`` type in the android implementation. Dictionaries are error prone.

aleffert replied:
Oh I see. there is a type below. We should return one of those.

abhra-clarice replied:
OK.

aleffert commented on edXVideoLocker/OEXCourseInfoViewController.m:
```diff
@@ -0,0 +1,84 @@
+-(void)reachabilityDidChange:(NSNotification *)notification{
+    [super reachabilityDidChange:notification];
+    if (self.dataInterface.reachable) {
+        [self.webViewHelper loadWebViewWithURLString:self.initialURLString];
```
We only need to do this if the content is not already loaded.

abhra-clarice replied:
Done.

aleffert commented on edXVideoLocker/OEXCourseInfoViewController.m:
```diff
@@ -0,0 +1,84 @@
+    }
+    
+    if (enrollmentExists) {
+        [[OEXFlowErrorViewController sharedInstance] showErrorWithTitle:@"Enrollment Error" message:@"You are already enrolled in this course" onViewController:self.view shouldHide:YES];
```
If they're already enrolled, we should just take them to the main screen.

abhra-clarice replied:
Done. I have put an status message pop up there, saying "You are now enrolled to the course".

aleffert commented on edXVideoLocker/OEXCourseInfoViewController.m:
```diff
@@ -0,0 +1,84 @@
+}
+
+-(void)dealloc{
+    self.initialURLString = nil;
```
This is unnecessary. All instance variables are cleared by the runtime on dealloc.

abhra-clarice replied:
Done.

aleffert commented on edXVideoLocker/OEXEnrollmentConfig.h:
```diff
@@ -0,0 +1,21 @@
+
+@interface OEXEnrollmentConfig : NSObject
+
++ (instancetype)sharedEnrollmentConfig;
```
This does not need to be a singleton. We should try to have as few of those as possible.

abhra-clarice replied:
OK.

aleffert commented on edXVideoLocker/OEXEnrollmentConfig.m:
```diff
@@ -0,0 +1,57 @@
+    
+}
+
+@property (nonatomic, strong) NSDictionary *courseEnrollmentDictionary;
```
Instead of having a dictionary as an ivar and read it from it every time, we should have an ``initWithDictionary:`` method that takes a dictionary and sets actual properties.

abhra-clarice replied:
Done.

aleffert commented on edXVideoLocker/OEXFindCoursesBaseViewController.m:
```diff
@@ -0,0 +1,101 @@
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    
+    self.navigationController.navigationBarHidden = YES;
```
Why are we hiding the navigation bar?

Also, when hiding the navigation bar, you should do it in viewWillAppear: . Loading sometimes happens at a different time than view appearance.

abhra-clarice replied:
Actually we are not using the default navigation bar. Instead we are using an object of OEXCustomNavigationView. That is why we have hidden the navigation bar. I have moved the code hiding to viewWillAppear:.

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.h:
```diff
@@ -0,0 +1,16 @@
+
+@interface OEXFindCoursesViewController : OEXFindCoursesBaseViewController
+
+@property (nonatomic) BOOL fromRearTable;
```
This property is not used.

abhra-clarice replied:
Line removed.

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.m:
```diff
@@ -0,0 +1,150 @@
+
+@interface OEXFindCoursesViewController () <SWRevealViewControllerDelegate, MFMailComposeViewControllerDelegate>
+
+@property (weak, nonatomic) IBOutlet UIButton *overlayButton;
```
``IBOutlet``s should be marked as ``strong``. It was kind of useful back when ``viewDidUnload`` was called. But since it isn't as of iOS 5 (I think), making them ``weak`` will pretty much only cause unexpected behavior.

abhra-clarice replied:
Should we set the IBOutlet properties to nil in dealloc?

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.m:
```diff
@@ -0,0 +1,150 @@
+    self.view.userInteractionEnabled=NO;
+    self.overlayButton.hidden = NO;
+    [self.navigationController popToViewController:self animated:NO];
+    [UIView animateWithDuration:0.9 delay:0 options:0 animations:^{
```
.9 is a really long animation duration? I haven't looked at it, so maybe it's fine, but is that really what we want?

aleffert replied:
Since you don't need a ``completion``, ``delay``, or ``options``, you can just do:

```
[UIView animateWithDuration:0.9 delay:0 options:0 animations:^{
    self.overlayButton.alpha = 0.5;
}]
```

abhra-clarice replied:
The duration of 0.9 is in sync with other animations used in the project.
 I have changed the code to:
```
[UIView animateWithDuration:0.9 animations:^{
        self.overlayButton.alpha = 0.5;
}];
```

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.m:
```diff
@@ -0,0 +1,150 @@
+        }];
+
+        OEXAppDelegate *appDelegate = (OEXAppDelegate *)[[UIApplication sharedApplication] delegate];
+        if (appDelegate.pendingMailComposerLaunch) {
```
Is this logic copied from somewhere else? Seems like it should be encapsulated. Also seems like we should set things up so every single controller doesn't need to know about this?

abhra-clarice replied:
Yes, you are correct. This logic is common amongst OEXFrontCourseViewController, OEXMyVideosViewController and OEXFindCoursesViewController with small differences.

aleffert replied:
But it's going to be trouble factoring it out from here? I'd rather not have this code proliferating everywhere.

abhra-clarice replied:
Added a new class "OEXRevealContentViewController" from which OEXFrontCourseViewController, OEXMyVideosViewController and OEXFindCoursesBaseViewController are subclassed. OEXRevealContentViewController will handle the delegation of revealViewController and the overlayButton action.

abhra-clarice replied:
Added an extension of OEXRevealContentViewController in "OEXRevealContentViewController+Protected.h" to contain the overlayButton and its action.

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.m:
```diff
@@ -0,0 +1,150 @@
+
+-(void)dealloc{
+    self.overlayButton = nil;
+    self.findCoursesURLString = nil;
```
unneeded ``dealloc``

abhra-clarice replied:
Removed dealloc.

aleffert commented on edXVideoLocker/OEXFindCoursesWebViewHelper.h:
```diff
@@ -0,0 +1,24 @@
+@class OEXFindCoursesWebViewHelper;
+
+@protocol OEXFindCoursesWebViewHelperDelegate <NSObject>
+@optional
```
Avoid ``@optional`` It's basically a big sign that says to the compiler "don't warn me about bugs"

aleffert replied:
There's a way we can structure this, where we pass in the path we're interested in to the constructor, so that the creator tells it what URLs it should be interested in. That way we don't need to do things optionally and it can also be used in more general way. I'll try to write up a gist tomorrow to make that clearer.

abhra-clarice replied:
Please elaborate how we can implement the general way. For now, I have removed the ```@optional```

aleffert replied:
Hi @abhra-clarice I stuck a gist here that demonstrates this https://gist.github.com/ee159c73e087cfeadb38 

As you can see it, just moves the responsibility of actually detecting the URL to the owner, since that's the thing that knows what's really going on.

It also makes a few other minor changes that were necessary in order to make this work. I haven't had a chance to actually run it, so keep an eye out if you integrate it.

abhra-clarice replied:
Hi,
I have added your changes into the code. The code is working fine with the
following changes:

1. In OEXFindCoursesViewController, changed
*static NSString* const OEXFindCoursesCourseInfoPath = @"course_info"; *
to
*static NSString* const OEXFindCoursesCourseInfoPath = @"course_info/";*

2. In OEXCourseInfoViewController, changed
*static NSString* const OEXFindCoursesEnrollPath = @"enroll";*
to
*static NSString* const OEXFindCoursesEnrollPath = @"enroll/";*

This is needed as the method -[NSURL oex_hostlessPath] returns values
appended by a front-slash when querying with the URLs.


On Sat, Feb 14, 2015 at 4:26 AM, Akiva Leffert <notifications@github.com>
wrote:

> In edXVideoLocker/OEXFindCoursesWebViewHelper.h
> <https://github.com/edx/edx-app-ios/pull/38#discussion_r24704314>:
>
> > @@ -0,0 +1,24 @@
> > +//
> > +//  OEXFindCoursesWebViewHelper.h
> > +//  edXVideoLocker
> > +//
> > +//  Created by Abhradeep on 02/02/15.
> > +//  Copyright (c) 2015 edX. All rights reserved.
> > +//
> > +
> > +#import <Foundation/Foundation.h>
> > +
> > +@class OEXFindCoursesWebViewHelper;
> > +
> > +@protocol OEXFindCoursesWebViewHelperDelegate <NSObject>
> > +@optional
>
> Hi @abhra-clarice <https://github.com/abhra-clarice> I stuck a gist here
> that demonstrates this https://gist.github.com/ee159c73e087cfeadb38
>
> As you can see it, just moves the responsibility of actually detecting the
> URL to the owner, since that's the thing that knows what's really going on.
>
> It also makes a few other minor changes that were necessary in order to
> make this work. I haven't had a chance to actually run it, so keep an eye
> out if you integrate it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-app-ios/pull/38/files#r24704314>.
>



-- 
Thanks & Regards,
Abhradeep Mazumder
Clarice Technologies
www.claricetechnologies.com

aleffert commented on edXVideoLocker/OEXFindCoursesWebViewHelper.m:
```diff
@@ -0,0 +1,115 @@
+    return YES;
+}
+
+- (void)webViewDidStartLoad:(UIWebView *)webView{
```
These methods are ``@optional`` so we don't need to implement them.

abhra-clarice replied:
Removed

aleffert commented on edXVideoLocker/OEXFindCoursesWebViewHelper.h:
```diff
@@ -0,0 +1,24 @@
+
+-(instancetype)initWithWebView:(UIWebView *)webView delegate:(id <OEXFindCoursesWebViewHelperDelegate>)delegate;
+
+-(void)loadWebViewWithURLString:(NSString *)urlString;
```
If you make the webView an argument to this method, then you don't need to have it as an instance variable, which will make this class simpler. The owner should have a reference to it already anyway.

abhra-clarice replied:
I have taken the webView as an argument and kept it in a weak-referenced property since methods such as loadWebViewWithURLString: are actively using it.

aleffert commented on edXVideoLocker/OEXFindCoursesWebViewHelper.m:
```diff
@@ -0,0 +1,115 @@
+#import "OEXEnrollmentConfig.h"
+
+#define OEXCourseInfoLinkURLPrefix @"edxapp://course_info?path_id="
+#define OEXCourseInfoLinkURLAlternatePrefix @"edxapp://view_course/course_path=course/"
```
Please don't use ``#defines``, the compiler doesn't know as much about them as it does about actual symbol definitions.

e.g.

static NSString* const OEXCourseInfoLinkURLPrefix = "edxapp://course_info?path_id="

aleffert replied:
I believe this string got updated to match what is in the spec.

abhra-clarice replied:
Looks like the URL has been updated to 
"edxapp://course_info?path_id=course/foobar"
instead of 
"edxapp://course_info?path_id=foobar" as mentioned in architecture document.

aleffert commented on edXVideoLocker/OEXFindCoursesWebViewHelper.m:
```diff
@@ -0,0 +1,115 @@
+#import "OEXFindCoursesWebViewHelper.h"
+#import "OEXEnrollmentConfig.h"
+
+#define OEXCourseInfoLinkURLPrefix @"edxapp://course_info?path_id="
```
Don't check for a prefix. Instead, parse the URL using NSURL and examine the components. This way if we add another argument the order of arguments won't matter. To get the query string parameters you can use this cocoa pod http://cocoadocs.org/docsets/NSURL+QueryDictionary/1.0.3/ . NSURL has a ``queryItems`` property, but you shouldn't use it since it's only on iOS 8.

abhra-clarice replied:
I have examined the scheme to be "edxapp" first and then went on to check the host which could either be "course_info" or "enroll". Depending on the host, I am then extracting the query parameters (e.g. "path_id", "course_id" and "email_opt_in"). The parsing of the query in done using the following method:
```
-(NSDictionary *)queryDictionaryForURL:(NSURL *)url{
    NSString *queryString = [url query];
    NSMutableDictionary *queryDictionary = [[NSMutableDictionary alloc] init];
    for (NSString *param in [queryString componentsSeparatedByString:@"&"]) {
        NSArray *keyValuePair = [param componentsSeparatedByString:@"="];
        if([keyValuePair count] < 2){
            continue;
        }
        [queryDictionary setObject:[keyValuePair objectAtIndex:1] forKey:[keyValuePair objectAtIndex:0]];
    }
    return queryDictionary;
}
```

aleffert commented on edXVideoLocker/OEXFindCoursesWebViewHelper.m:
```diff
@@ -0,0 +1,115 @@
+- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType{
+    if (navigationType ==  UIWebViewNavigationTypeLinkClicked) {
+        NSString *linkURLPrefixString = nil;
+        if ([[request.URL absoluteString] hasPrefix:OEXCourseInfoLinkURLPrefix]) {
```
As discussed all of this code should be working with the parts of the URL instead of the absoluteString. That will be a lot more robust.

abhra-clarice replied:
Removed checks dependent on absoluteString.

aleffert commented on edXVideoLocker/OEXFindCoursesWebViewHelper.m:
```diff
@@ -0,0 +1,115 @@
+}
+
+-(void)dealloc{
+    self.webView = nil;
```
unnecessary ``dealloc``

abhra-clarice replied:
Removed dealloc.

aleffert commented on edXVideoLocker/OEXRearTableViewController.m:
```diff
@@ -82,6 +84,10 @@ - (void) prepareForSegue: (UIStoryboardSegue *) segue sender: (id) sender
         {
             UINavigationController* nc = [[UINavigationController alloc] initWithRootViewController:dvc];
             [rvc pushFrontViewController:nc animated:YES];
+            if ([sender isKindOfClass:[OEXCourseInfoViewController class]] && [dvc isKindOfClass:[OEXFrontCourseViewController class]]) {
```
This view shouldn't be responsible for forwarding that message. They have nothing to do with each other.

Also, we should avoid doing ``isKindOfClass:`` checks on view controllers. We have longer term plans to let third parties customize which classes are used for which screens.

 If we really need the front view controller to be responsible for showing it, I think this is a good use for NSNotification.

abhra-clarice replied:
Used NSNotification.

aleffert commented on edXVideoLocker/OEXURLSessionManager.h:
```diff
@@ -0,0 +1,17 @@
+
+@interface OEXURLSessionManager : NSObject
+
++ (instancetype)sharedURLSessionManager;
```
We don't need a separate class with a singleton just for one method. It seems like this method should be added onto the ``OEXNetworkManager``?

abhra-clarice replied:
I have moved the method to OEXNetworkManager. The URL session currently used in the method is obtained by calling sessionForRequest: of OEXNetworkManager

aleffert commented on edXVideoLocker/OEXURLSessionManager.m:
```diff
@@ -0,0 +1,41 @@
+    
+    [request setHTTPBody:body];
+    
+    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration] delegate:nil delegateQueue:[NSOperationQueue mainQueue]];
```
This makes a new URL session every time it's called, which is probably not what we want.

aleffert commented on edXVideoLocker/OEXURLSessionManager.m:
```diff
@@ -0,0 +1,41 @@
+    return sharedURLSessionManager;
+}
+
+-(void)callWebServiceWithURLPath:(NSString *)urlPath method:(NSString *)method body:(NSData *)body completionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))completionHandle{
```
It's worth putting the word "authorized" in the name for clarity.

abhra-clarice replied:
Changed the method name to:
```
-(void)callAuthorizedWebServiceWithURLPath:(NSString *)urlPath method:(NSString *)method body:(NSData *)body completionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))completionHandle;
```

aleffert commented on edXVideoLocker/OEXURLSessionManager.m:
```diff
@@ -0,0 +1,41 @@
+    [request setHTTPMethod:method];
+    
+    NSString *authValue = [NSString stringWithFormat:@"%@",[OEXAuthentication authHeaderForApiAccess]];
+    [request setValue:authValue forHTTPHeaderField:@"Authorization"];
```
I like standardizing a path for authorized requests. It doesn't need to be part of this pull request, but we should migrate existing calls to use one path like this.

aleffert commented on edXVideoLocker/OEXCourseInfoViewController.m:
```diff
@@ -0,0 +1,84 @@
+    
+    NSData *enrollmentJSONData = [NSJSONSerialization dataWithJSONObject:enrollmentDictionary options:0 error:nil];
+    
+    [urlSessionManager callWebServiceWithURLPath:URL_COURSE_ENROLLMENT method:@"POST" body:enrollmentJSONData completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
```
Instead of writing @"POST" everywhere, which is error prone, let's define some string constants for these. 

```
NSString* const OEXHTTPMethodPOST = @"POST"
``
or something like that.

abhra-clarice replied:
I have defined a new constant in OEXConstants.h:
``` #define OEXHTTPMethodPOST @"POST" ```

aleffert commented on edXVideoLocker/OEXFindCoursesBaseViewController.h:
```diff
@@ -0,0 +1,35 @@
+
+@property (weak, nonatomic) IBOutlet DACircularProgressView *customProgressBar;
+@property (weak, nonatomic) IBOutlet OEXCustomNavigationView *customNavView;
+@property (weak, nonatomic) IBOutlet UIButton *showDownloadsButton;
```
The works with subclassing, we have to expose a bunch of properties and methods that shouldn't be part of the interface. I'd be happier with a more compositional approach, but if that's going to be a pain, can you move all the stuff that only subclasses should see to a separate header and call it ``OEXFindCoursesBaseViewController+Protected.h"

abhra-clarice replied:
This has been implemented using an anonymous category on OEXFindCoursesBaseViewController as a named category would not let us store properties in memory.

aleffert commented on edXVideoLocker/OEXFindCoursesBaseViewController.m:
```diff
@@ -0,0 +1,101 @@
+- (void)hideOfflineLabel:(BOOL)isOnline{
+    self.customNavView.lbl_Offline.hidden = isOnline;
+    self.customNavView.view_Offline.hidden = isOnline;
+//    [self.customNavView adjustPositionIfOnline:isOnline];
```
Can we remove this commented out line?

abhra-clarice replied:
Sure.

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.m:
```diff
@@ -0,0 +1,150 @@
+
+-(void)reachabilityDidChange:(NSNotification *)notification{
+    [super reachabilityDidChange:notification];
+    if (self.dataInterface.reachable) {
```
Again, we should only do this if the content isn't already loaded.

abhra-clarice replied:
OK.

aleffert commented:
This needs tests.

abhra-clarice commented:
* Implemented the comments.
* Added scenario where the "enabled" flag is false.
* Added a test case class "OEXFindCoursesTests"

@aleffert , Please Review.

aleffert commented on edXVideoLocker/OEXCourseInfoViewController.m:
```diff
@@ -0,0 +1,99 @@
+    }
+    
+    if (enrollmentExists) {
+        [self performSelectorOnMainThread:@selector(showMainScreenWithMessage:) withObject:@"You are already enrolled to this course" waitUntilDone:NO];
```
Instead of performSelector, use dispatch_async. That way the compiler can type check the call.
For example,

```
dispatch_async(dispatch_get_main_queue()) {
  [self showMainScreenWithMessage:...]
});
```

Also, this string needs to be localized.

abhra-clarice replied:
Removed performSelector, since this line is always executed in main thread itself.

aleffert commented on edXVideoLocker/OEXCourseInfoViewController.m:
```diff
@@ -0,0 +1,99 @@
+        NSLog(@"Res: %@ data: %@",response, [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]);
+        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
+        if (httpResponse.statusCode == 200) {
+            [self performSelectorOnMainThread:@selector(showMainScreenWithMessage:) withObject:@"You are now enrolled to the course" waitUntilDone:NO];
```
Same here.

abhra-clarice replied:
done.

aleffert commented on edXVideoLocker/OEXCourseInfoViewController.m:
```diff
@@ -0,0 +1,99 @@
+}
+
+-(void)showEnrollmentError{
+    [[OEXFlowErrorViewController sharedInstance] showErrorWithTitle:@"Enrollment Error" message:@"An error occurred while creating the new course enrollment" onViewController:self.view shouldHide:YES];
```
These also need to be localized

abhra-clarice replied:
Done.

aleffert commented on edXVideoLocker/OEXEnrollmentConfig.h:
```diff
@@ -0,0 +1,20 @@
+
+@interface OEXEnrollmentConfig : NSObject
+
+-(instancetype)initWithDictionary:(NSDictionary *)dictionary;
```
I think we can make these readonly in the interface (and then in the ``.m`` file add a class continuation (category with an empty name) that redeclares them ``readwrite``. This way they won't be mutated unexpectedly.

abhra-clarice replied:
Done.

aleffert commented on edXVideoLocker/OEXFindCoursesBaseViewController.m:
```diff
@@ -0,0 +1,110 @@
+
+#define kShouldShowDownloadProgress NO
+
+@interface OEXFindCoursesBaseViewController ()
```
This declaration is empty so we don't need it.

abhra-clarice replied:
Lines removed.

aleffert commented on edXVideoLocker/OEXFindCoursesBaseViewController.m:
```diff
@@ -0,0 +1,110 @@
+    self.dataInterface = [OEXInterface sharedInterface];
+    [self setNavigationBar];
+    
+    if (kShouldShowDownloadProgress) {
```
Since ``kShouldShowDownloadProgress`` is hardcoded to NO should we just delete the other case?

abhra-clarice replied:
Sure.

aleffert commented on edXVideoLocker/OEXFindCoursesBaseViewController.m:
```diff
@@ -0,0 +1,110 @@
+}
+
+-(void)dealloc{
+    self.customProgressBar = nil;
```
All properties are set to nil automatically (including weak ones) when an object gets dealloc'd so we don't need any of this.

abhra-clarice replied:
dealloc removed.

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.m:
```diff
@@ -0,0 +1,166 @@
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    
+    self.findCoursesURLString = [[[OEXConfig sharedConfig] courseEnrollmentConfig] searchURL];
```
We don't need an instance variable to store this. It just creates unnecessary state. This object could just have a method that returns the one from the config.

aleffert replied:
How about just doing,
```
- (OEXEnrollmentConfig*)enrollmentConfig {
    return [[OEXConfig sharedConfig] courseEnrollmentConfig];
}
```

and then you can just do [self enrollmentConfig].searchURL or whatever.

abhra-clarice replied:
Code modified.

aleffert commented on edXVideoLocker/OEXFindCoursesViewController.m:
```diff
@@ -0,0 +1,166 @@
+    
+    self.overlayButton.alpha = 0.0f;
+    
+    if (![[[OEXConfig sharedConfig] courseEnrollmentConfig] enabled]) {
```
and also can use it here.

abhra-clarice replied:
Sure.

aleffert commented on edXVideoLockerTests/OEXFindCoursesTests.m:
```diff
@@ -0,0 +1,78 @@
+}
+
+-(void)testEnrollmentConfig{
+    OEXEnvironment *environment = [[OEXEnvironment alloc] init];
```
Instead of creating an environment, which sets a bunch of globals, for testing, you can just instantiate an OEXConfig directly, using a dictionary. This way, the test doesn't depend on the particular config the person running the tests is using.

abhra-clarice replied:
Initialized ```config``` as ```OEXConfig *config = [[OEXConfig alloc] initWithAppBundleData];```

aleffert commented on edXVideoLockerTests/OEXFindCoursesTests.m:
```diff
@@ -0,0 +1,78 @@
+    
+    OEXNetworkManager *sharedManager = [OEXNetworkManager sharedManager];
+
+    NSDictionary *enrollmentDictionary = @{@"course_details":@{@"course_id": @"course-v1:BerkeleyX+GG101x-2+1T2015", @"email_opt_in":@"false"}};
```
This test talks to our live server. Running the unit tests shouldn't require an internet connection. We should be mocking responses for things like that. OHHTTPStubs is a good mocking library for web servers.

abhra-clarice replied:
Removed ```testCallAuthorizedWebService```

aleffert commented on edXVideoLockerTests/OEXFindCoursesTests.m:
```diff
@@ -0,0 +1,78 @@
+    [self waitForExpectationsWithTimeout:10 handler:^(NSError *error) {
+
+    }];
+}
```
We should also have some tests that verify that we're correctly recognizing URLs and that we're doing template substitution correctly.

abhra-clarice replied:
Updated OEXFindCoursesTests.m

abhra-clarice commented:
1. Implemented the comments
2. Updated the test cases.

@aleffert, Please review.

abhra-clarice commented:
1. Updated test cases.
2. Added class for Enrollment notification message to determine whether to reload the course list table or not.

@aleffert , Please review.

aleffert commented:
Looks good.
